### PR TITLE
!feat: Add invartiants

### PR DESCRIPTION
**Pull request by GNaD13**
This PR should be merge after #111

### CODE REVIEW

File: x/multi-staking/keeper/invariants.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+		var multiStakingUnlocks []types.MultiStakingUnlock
+		k.MultiStakingUnlockIterator(ctx, func(stakingUnlock types.MultiStakingUnlock) bool {
+			multiStakingUnlocks = append(multiStakingUnlocks, stakingUnlock)
+			return false
+		})
+
+		for _, unlock := range multiStakingUnlocks {
+			valAddr := unlock.UnlockID.ValAddr
+			valAllowedDenom := k.GetValidatorMultiStakingCoin(ctx, sdk.ValAddress(valAddr))
```
Comment (neitdung) [inline x/multi-staking/keeper/invariants.go:89]: We should change variable name to `valMsDenom`

File: x/multi-staking/keeper/abci.go [hunk 2]
```diff
@@ -87,6 +87,19 @@ func (k Keeper) BurnUnbondedCoinAndUnlockedMultiStakingCoin(
...
 
+	// burn slash coin
+	slashAmount := unlockEntry.UnlockingCoin.Amount.Sub(unlockedAmount)
+	slashCoin := sdk.NewCoin(unlockDenom, slashAmount)
+	err = k.bankKeeper.BurnCoins(ctx, types.ModuleName, sdk.NewCoins(slashCoin))
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	err = k.UnescrowCoinTo(ctx, multiStakerAddr, slashCoin)
```
[the remark this answers is not in the document]
Comment (GNaD13) [reply x/multi-staking/keeper/abci.go]: this is old code for burn the slash token, not use this approach anymore. I've deleted it

File: x/multi-staking/module.go [hunk 3]
```diff
@@ -109,6 +109,8 @@ func (AppModule) Name() string {
 // RegisterInvariants registers the staking module invariants.
 // TODO: Need to implement invariants
```
[the remark this answers is not in the document]
Comment (GNaD13) [reply x/multi-staking/module.go]: done

File: x/multi-staking/keeper/invariants.go (hunk 1 above)
Comment (GNaD13) [reply x/multi-staking/keeper/invariants.go]: tysm

File: x/multi-staking/keeper/invariants.go [hunk 4]
```diff
@@ -73,7 +73,7 @@ func ValidatorLockDenomInvariants(k Keeper) sdk.Invariant {
 			if valMsDenom := k.GetValidatorMultiStakingCoin(ctx, sdk.ValAddress(valAddr)); valMsDenom != lock.LockedCoin.Denom {
 				broken = true
 				msg += fmt.Sprintf("validator lock denom invariants:\n\t lock denom: %v"+
-					"\n\tvalidator allow denom: %v\n",
+					"\n\tvalidator valMsDenom: %v\n",
```
Comment (neitdung) [inline x/multi-staking/keeper/invariants.go:76]: This log should be validator denom or origin validator allow denom

**Comment by catShaark** [comment]
Can you add some tests

**Comment by GNaD13** [comment]
@catShaark could you check this pr pls?

File: x/multi-staking/keeper/invariants.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+			lockCoinAmount = lockCoinAmount.Add(sdk.NewCoin(stakingLock.LockedCoin.Denom, stakingLock.LockedCoin.Amount))
+			return false
+		})
+		totalLockCoinAmount = totalLockCoinAmount.Add(lockCoinAmount...)
+
+		// calculate unlocking amount
+		unlockingCoinAmount := sdk.NewCoins()
+		k.MultiStakingUnlockIterator(ctx, func(unlock types.MultiStakingUnlock) bool {
+			for _, entry := range unlock.Entries {
+				unlockingCoinAmount = unlockingCoinAmount.Add(sdk.NewCoin(entry.UnlockingCoin.Denom, entry.UnlockingCoin.Amount))
```
Comment (catShaark) [inline x/multi-staking/keeper/invariants.go:36]: can you use .ToCoin()

File: x/multi-staking/keeper/invariants.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+func ModuleAccountInvariants(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		totalLockCoinAmount := sdk.NewCoins()
+
+		// calculate lock amount
+		lockCoinAmount := sdk.NewCoins()
+		k.MultiStakingLockIterator(ctx, func(stakingLock types.MultiStakingLock) bool {
+			lockCoinAmount = lockCoinAmount.Add(sdk.NewCoin(stakingLock.LockedCoin.Denom, stakingLock.LockedCoin.Amount))
```
Comment (catShaark) [inline x/multi-staking/keeper/invariants.go:27]: can you use .ToCoin()

File: x/multi-staking/keeper/invartiants_test.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+			DelegatorAddress:  delAddr.String(),
+			ValidatorAddress:  valAddr.String(),
+			Pubkey:            codectypes.UnsafePackAny(valPubKey),
+			Value:             bondAmount,
+		}
+
+		_, err = suite.msgServer.CreateValidator(suite.ctx, &msg)
+		suite.Require().NoError(err)
+
+		tc.malleate()
```
Comment (catShaark) [inline x/multi-staking/keeper/invartiants_test.go:154]: why there's no fail case in this test

File: x/multi-staking/keeper/store_test.go [hunk 8]
```diff
@@ -86,3 +89,43 @@ func (suite *KeeperTestSuite) TestSetValidatorMultiStakingCoin() {
...
+			Amount:     sdk.NewIntFromUint64(1000000),
+			BondWeight: sdk.NewDec(1),
+		},
+	}
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError bool
+	}{
```
Comment (catShaark) [inline x/multi-staking/keeper/store_test.go:114]: there's no fail case for this test

File: x/multi-staking/keeper/invartiants_test.go (hunk 7 above)
Comment (GNaD13) [reply x/multi-staking/keeper/invartiants_test.go:154]: invariants should be success in any case

File: x/multi-staking/keeper/store_test.go (hunk 8 above)
Comment (GNaD13) [reply x/multi-staking/keeper/store_test.go:114]: This is different test for check the store, not related in invariants. We should continue this test in different PR

File: x/multi-staking/keeper/invartiants_test.go (hunk 7 above)
Comment (catShaark) [reply x/multi-staking/keeper/invartiants_test.go:154]: there should be fail case in test, just set data in the store that causes it to fail

File: x/multi-staking/keeper/store_test.go (hunk 8 above)
Comment (catShaark) [reply x/multi-staking/keeper/store_test.go:114]: there should be fail case anyway because there is tc.expError.

Comment (GNaD13) [reply x/multi-staking/keeper/store_test.go:114]: If so I think we should move it to a separate PR

Comment (catShaark) [reply x/multi-staking/keeper/store_test.go:114]: it's okay to put the test in this pr

**Pull request merged**